Dart/src/com/jetbrains/lang/dart/ide/editor/DartEnterInStringHandler.java

```diff
@@ -0,0 +1,56 @@
+package com.jetbrains.lang.dart.ide.editor;
+
+import com.intellij.codeInsight.editorActions.enter.EnterHandlerDelegateAdapter;
+import com.intellij.lang.ASTNode;
+import com.intellij.openapi.actionSystem.DataContext;
+import com.intellij.openapi.editor.Document;
+import com.intellij.openapi.editor.Editor;
+import com.intellij.openapi.editor.actionSystem.EditorActionHandler;
+import com.intellij.openapi.util.Ref;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.tree.IElementType;
+import com.jetbrains.lang.dart.DartTokenTypes;
+import org.jetbrains.annotations.NotNull;
+
+public class DartEnterInStringHandler extends EnterHandlerDelegateAdapter {
+
+  @Override
+  public Result preprocessEnter(@NotNull final PsiFile file,
+                                @NotNull final Editor editor,
+                                @NotNull final Ref<Integer> caretOffsetRef,
+                                @NotNull final Ref<Integer> caretAdvanceRef,
+                                @NotNull final DataContext dataContext,
+                                final EditorActionHandler originalHandler) {
+    int caretOffset = caretOffsetRef.get().intValue();
+    int caretAdvance = caretAdvanceRef.get().intValue();
+    PsiElement psiAtOffset = file.findElementAt(caretOffset);
+    if (psiAtOffset == null || psiAtOffset.getTextOffset() > caretOffset) {
+      return Result.Continue;
+    }
+    Document document = editor.getDocument();
+    ASTNode token = psiAtOffset.getNode();
+    IElementType type = token.getElementType();
+    if (type == DartTokenTypes.REGULAR_STRING_PART) {
+      token = token.getTreeNext();
+      type = token.getElementType();
```

Comment:
Well, you are right, may be in this case null can't appear. Personally I prefer to add an extra null-check rather than think about data flow and hope that everyone who will be updating lexer/parser/DartEnterInStringHandler in future won't forget to add null-check when null become real.